server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -347,4 +357,113 @@ public String toString() {
+        private final NodeConnectionHistory dummy = new NodeConnectionHistory("", 0, null);
+        private final ConcurrentMap<String, NodeConnectionHistory> nodeHistory = ConcurrentCollections.newConcurrentMap();
+
+        ConnectionHistory() {
+            NodeConnectionsService.this.transportService.addConnectionListener(new TransportConnectionListener() {
+                @Override
+                public void onNodeConnected(DiscoveryNode node, Transport.Connection connection) {
+                    // log case where the remote node has same ephemeralId as its previous connection
+                    // (the network was disrupted, but not the remote process)
+                    NodeConnectionHistory nodeConnectionHistory = nodeHistory.get(node.getId());
+                    if (nodeConnectionHistory != null) {
+                        nodeHistory.replace(node.getId(), nodeConnectionHistory, dummy);
+                    }
+
+                    if (nodeConnectionHistory != null
+                        && nodeConnectionHistory != dummy
+                        && nodeConnectionHistory.ephemeralId.equals(node.getEphemeralId())) {
+                        if (nodeConnectionHistory.disconnectCause != null) {
+                            logger.warn(
+                                () -> format(
+                                    "reopened transport connection to node [%s] "
+                                        + "which disconnected exceptionally [%dms] ago but did not "
+                                        + "restart, so the disconnection is unexpected; "
+                                        + "if unexpected, see [{}] for troubleshooting guidance",
+                                    node.descriptionWithoutAttributes(),
+                                    nodeConnectionHistory.disconnectTime,
+                                    ReferenceDocs.NETWORK_DISCONNECT_TROUBLESHOOTING
+                                ),
+                                nodeConnectionHistory.disconnectCause
+                            );
+                        } else {
+                            logger.warn(
+                                """
+                                    reopened transport connection to node [{}] \
+                                    which disconnected gracefully [{}ms] ago but did not \
+                                    restart, so the disconnection is unexpected; \
+                                    if unexpected, see [{}] for troubleshooting guidance""",
+                                node.descriptionWithoutAttributes(),
+                                nodeConnectionHistory.disconnectTime,
+                                ReferenceDocs.NETWORK_DISCONNECT_TROUBLESHOOTING
+                            );
+                        }
+                    }
+                }
+
+                @Override
+                public void onNodeDisconnected(DiscoveryNode node, Transport.Connection connection) {
```

Comment:
I just spotted we're already executing this in a close-listener, but one that runs under `ActionListener.running(...)` so it drops the exception. I think it'd be nicer to adjust this callback to take a `@Nullable Exception e` parameter rather than having to add a second close listener just to pick up the exception as done here.